watchtracker/watching_tracker.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2019-2020 VMware, Inc.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package watchtracker
+
+import (
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/vmware-labs/reconciler-runtime/reconcilers"
+	"github.com/vmware-labs/reconciler-runtime/tracker"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// WatchingTracker defines an interface through which an object can register
+// that it is tracking another object by reference and which will automatically
+// establish an informer watch for the group, version, and kind of the tracked
+// object.
+type WatchingTracker interface {
+	// Track tells us that "obj" is tracking changes to the referenced object
+	// with the given version and establishes and informer watch for the group
+	// and kind of the referenced object and the given version.
+	Track(ref tracker.Key, version string, obj types.NamespacedName) error
+
+	// Lookup returns actively tracked objects for the reference.
+	Lookup(ref tracker.Key) []types.NamespacedName
+}
+
+type impl struct {
+	tracker    tracker.Tracker
+	controller controller.Controller
+	scheme     *runtime.Scheme
+
+	m       sync.Mutex // protects watches
+	watches map[schema.GroupVersionKind]struct{}
+}
+
+// New returns an implementation of WatchingTracker that lets a Reconciler
+// register a particular resource as watching a resource for
+// a particular lease duration.  This watch must be refreshed
+// periodically (e.g. by a controller resync) or it will expire.
+func New(lease time.Duration, log logr.Logger, controller controller.Controller, scheme *runtime.Scheme) WatchingTracker {
+	return newWatchingTracker(tracker.New(lease, log), controller, scheme)
+}
+
+func newWatchingTracker(tracker tracker.Tracker, controller controller.Controller, scheme *runtime.Scheme) WatchingTracker {
+	return &impl{
+		tracker:    tracker,
+		controller: controller,
+		scheme:     scheme,
+		watches:    map[schema.GroupVersionKind]struct{}{},
```

Comment:
No need to index with the version as it's the group and kind that are important. We still need a GVK to setup the watch, but not to flag that a watch has be started.

```suggestion
		watches:    map[schema.GroupKind]struct{}{},
```

Reply:
> No need to index with the version as it's the group and kind that are important. We still need a GVK to setup the watch, but not to flag that a watch has be started.

There are two issues here:
1. It is not clear from the controller runtime interfaces that two watches which differ only in version are identical in behaviour. This may well be the case, but if so, I'd like to see it documented in controller runtime, so we can really depend on it.
2. Also, and related to the first issue, it depends what interface controller runtime eventually provides for stopping watches. If that takes a version, then we'll need to record the version.

Reply:
> It is not clear from the controller runtime interfaces that two watches which differ only in version are identical in behaviour. This may well be the case, but if so, I'd like to see it documented in controller runtime, so we can really depend on it.

You won't find this documented, because they are not equivalent informers. However, because we only care about the metadata for the resource and not its content, they are equivalent to us.

> Also, and related to the first issue, it depends what interface controller runtime eventually provides for stopping watches. If that takes a version, then we'll need to record the version.

Let's cross that bridge when we get to it. Worse case, we look at the content of the informer and tease out the version.

I'm only suggesting that that map key use a GroupKind without the version to prevent setting up duplicate informers for different versions of the same resource that would be 100% overhead and provide no actual value.